### PR TITLE
TAJO-1306: HAServiceUtil should not directly use HDFS.

### PR DESCRIPTION
**Pull request by hyunsik**
In this patch, I added an abstract class ServiceTracker which registers Tajo rpc services and provides discovery services for them. I changed  almost all parts using HAServiceUtil to use ServiceTracker.

### CODE REVIEW

File: tajo-core/src/main/java/org/apache/tajo/ha/HdfsServiceTracker.java [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+    sb.append(address.getAddress().getHostAddress()).append(":").append(address.getPort()).append("_");
+
+    address = getHostAddress(HAConstants.MASTER_INFO_ADDRESS);
+    sb.append(address.getAddress().getHostAddress()).append(":").append(address.getPort());
+
+    FSDataOutputStream out = fs.create(path);
+
+    try {
+      out.writeUTF(sb.toString());
+      out.hflush();
```
Comment (jinossy) [inline tajo-core/src/main/java/org/apache/tajo/ha/HdfsServiceTracker.java]: hflush() is client side buffer flushing. Can you change to hsync() ?

**Comment by blrunner** [comment]
Hi @hyunsik 

Thanks for your contribution and simple codes by ServiceTracker looks very pretty to me.
But unfortunately, I couldn't run Tajo HA cluster with a shell script (start-tajo.sh) on my macbook as follows:

2015-01-23 11:15:46,876 INFO org.apache.tajo.master.rm.TajoWorkerResourceManager: Released QueryMaster (q_1421979321683_0001) resource.
2015-01-23 11:15:46,885 INFO org.apache.tajo.util.history.HistoryWriter: Create query history file: hdfs://localhost:9010/tmp/tajo-blrunner/staging/history/20150123/query-list/query-list-111546.hist
2015-01-23 11:15:57,739 INFO org.apache.tajo.session.SessionManager: Session f103d278-ab81-47f4-a19b-3e9f561d73fd is removed.
2015-01-23 11:16:09,009 ERROR org.apache.tajo.master.TajoMaster: RECEIVED SIGNAL 15: SIGTERM
2015-01-23 11:16:09,010 INFO org.apache.tajo.master.TajoMaster: ============================================
2015-01-23 11:16:09,010 INFO org.apache.tajo.master.TajoMaster: TajoMaster received SIGINT Signal
2015-01-23 11:16:09,010 INFO org.apache.tajo.master.TajoMaster: ============================================
2015-01-23 11:16:09,013 INFO org.apache.tajo.catalog.CatalogServer: Catalog Server (127.0.0.1:26005) shutdown
2015-01-23 11:16:09,020 INFO org.apache.tajo.rpc.NettyServerBase: Rpc (CatalogProtocol) listened on 127.0.0.1:26005) shutdown
2015-01-23 11:16:09,021 INFO org.apache.tajo.catalog.store.MySQLStore: Shutdown database (jdbc:mysql://localhost:3306/tajo?createDatabaseIfNotExist=true)
2015-01-23 11:16:09,022 INFO org.apache.tajo.util.history.HistoryWriter: HistoryWriter_127.0.0.1_26001 stopped.
2015-01-23 11:16:09,022 INFO org.apache.tajo.util.history.HistoryCleaner: History cleaner stopped
2015-01-23 11:16:09,023 INFO org.apache.tajo.rpc.NettyServerBase: Rpc (QueryCoordinatorProtocol) listened on 127.0.0.1:26001) shutdown
2015-01-23 11:16:09,025 INFO org.apache.tajo.rpc.NettyServerBase: Rpc (TajoMasterClientProtocol) listened on 127.0.0.1:26002) shutdown
2015-01-23 11:16:09,026 INFO org.apache.hadoop.yarn.util.AbstractLivelinessMonitor: SessionLivelinessMonitor thread interrupted
2015-01-23 11:16:09,026 ERROR org.apache.tajo.master.rm.TajoWorkerResourceManager: java.lang.InterruptedException
2015-01-23 11:16:09,027 INFO org.apache.tajo.rpc.NettyServerBase: Rpc (TajoResourceTrackerProtocol) listened on 127.0.0.1:26003) shutdown
2015-01-23 11:16:09,027 INFO org.apache.hadoop.yarn.util.AbstractLivelinessMonitor: WorkerLivelinessMonitor thread interrupted
2015-01-23 11:16:09,028 INFO org.apache.tajo.master.TajoMaster: Tajo Master main thread exiting
2015-01-23 11:16:09,117 INFO org.apache.tajo.master.TajoMaster: SHUTDOWN_MSG:
/************************************************************
SHUTDOWN_MSG: Shutting down TajoMaster at jaehwaui-mbp.server.gruter.com/192.168.0.245
************************************************************/

For reference, I could run Tajo HA cluster with Tajo command as follows:
$TAJO_HOME/bin/tajo master

Cheers
Jaehwa

**Comment by blrunner** [comment]
For reference, after finding your comment, I'll check the patch such as JDBC, HA on YARN, auto-failover.

**Comment by hyunsik** [comment]
Thank you for your testing. I'll dig into the problem :)

**Comment by hyunsik** [comment]
@blrunner TajoMaster seems to be shutdown according to your log.

Could you point out any abnormal line in the log? I could find only SIGTERM signal. It seems to be triggered from external something.

**Comment by blrunner** [comment]
Sorry @hyunsik .

I found that SIGTERM signal had happened by wrong configuration. 
I'm testing the patch with various environment.
I'll give feedback as soon as possible. :)

**Comment by hyunsik** [comment]
I just fixed TajoWorkerClientService::getQueryHistory bug and triggered TravisCI.

**Comment by blrunner** [comment]
+1

It looks good to me and the patch ran as expected on my testing cluster.
Push it. :)

**Comment by blrunner** [comment]
Hi @hyunsik 

The patch ran as expected on my testing cluster and all unit test cases finished successfully.
But Travis CI build failed as follows:

{code:xml}
Running org.apache.tajo.ha.TestHAServiceHDFSImpl
Formatting using clusterid: testClusterID
2015-01-26 08:19:56,323 INFO: org.mortbay.log (info(67)) - jetty-6.1.14
2015-01-26 08:19:56,329 INFO: org.mortbay.log (info(67)) - Extract jar:file:/home/travis/.m2/repository/org/apache/hadoop/hadoop-hdfs/2.5.1/hadoop-hdfs-2.5.1-tests.jar!/webapps/hdfs to /tmp/Jetty_localhost_60866_hdfs____ofqbhk/webapp
2015-01-26 08:19:56,472 WARN: org.apache.hadoop.security.authentication.server.AuthenticationFilter (init(166)) - 'signature.secret' configuration not set, using a random value as secret
2015-01-26 08:19:56,474 INFO: org.mortbay.log (info(67)) - Started HttpServer2$SelectChannelConnectorWithSafeStartup@localhost:60866
2015-01-26 08:19:57,404 INFO: org.mortbay.log (info(67)) - jetty-6.1.14
2015-01-26 08:19:57,412 INFO: org.mortbay.log (info(67)) - Extract jar:file:/home/travis/.m2/repository/org/apache/hadoop/hadoop-hdfs/2.5.1/hadoop-hdfs-2.5.1-tests.jar!/webapps/datanode to /tmp/Jetty_localhost_43395_datanode____y4jjpz/webapp
2015-01-26 08:19:57,569 INFO: org.mortbay.log (info(67)) - Started HttpServer2$SelectChannelConnectorWithSafeStartup@localhost:43395
2015-01-26 08:19:58,281 WARN: org.apache.tajo.storage.FileStorageManager (storageInit(100)) - does not support block metadata. ('dfs.datanode.hdfs-blocks-metadata.enabled')
2015-01-26 08:20:01,009 ERROR: org.apache.tajo.master.rm.TajoWorkerResourceManager (run(346)) - java.lang.InterruptedException
2015-01-26 08:20:08,044 ERROR: org.apache.tajo.master.rm.TajoWorkerResourceManager (run(346)) - java.lang.InterruptedException
2015-01-26 08:20:09,734 ERROR: org.apache.tajo.worker.WorkerHeartbeatService (run(210)) - org.apache.tajo.service.ServiceTrackerException: No active master entry
org.apache.tajo.service.ServiceTrackerException: org.apache.tajo.service.ServiceTrackerException: No active master entry
	at org.apache.tajo.ha.HdfsServiceTracker.getAddressElements(HdfsServiceTracker.java:436)
	at org.apache.tajo.ha.HdfsServiceTracker.getResourceTrackerAddress(HdfsServiceTracker.java:361)
	at org.apache.tajo.worker.WorkerHeartbeatService$WorkerHeartbeatThread.run(WorkerHeartbeatService.java:188)
Caused by: org.apache.tajo.service.ServiceTrackerException: No active master entry
	at org.apache.tajo.ha.HdfsServiceTracker.getAddressElements(HdfsServiceTracker.java:408)
	... 2 more
{code}

**Comment by hyunsik** [comment]
Hi @blrunner ,

That is a part of original unit tests; i.e.g, TestHAServiceHDFSImpl::testAutoFailOver. The failure is not related to this unit test because the following unit tests are passes well until the failure point. According to logs, the failure seems to be caused by HDFS mini cluster, which occurs occasionally. Anyway, I trigger the TravisCI.

**Comment by blrunner** [comment]
+1

I think that the error cause was TAJO-1318.
Shit it. :)

**Pull request closed without merging**